catchup/service.go

```diff
@@ -475,7 +520,10 @@ func (s *Service) pipelinedFetch(seedLookback uint64) {
 			s.handleUnsupportedRound(nextRound)
 			break
 		}
-
+		// pause here until resume catchup sends a signal
+		if s.pauseAtRound != uint64(0) && uint64(nextRound) > s.pauseAtRound {
+			s.pauseAtRound = <-s.chanPauseAtRound
+		}
```

Comment:
I think it would be possible to solve this in a general way with a helper function with these changes:
1) initialize `chanPauseAtRound` once during the service initializer.
2) SetPauseAtRound changes:
```

// SetPauseAtRound changes the value of s.pauseAtRound
func (s *Service) SetPauseAtRound(rnd uint64) (err error) {
	// check if pause functionality is enabled
	if !s.dontAllowPauseAtRound {
		s.chanPauseAtRound <- rnd
	} else {
		err = errors.New("not allowed to pause")
	}
	return
}```
3) Helper function:
```
func (s *Service) updatePauseAtRoundOrBlock() {
	if s.pauseAtRound != uint64(0) && uint64(nextRound) > s.pauseAtRound {
		// if we have reached the desired round, block until the pause round changes
		s.pauseAtRound = <-s.chanPauseAtRound
	} else {
		// otherwise, just check if the pause round changed
		select {
			case s.pauseAtRound = <- s.chanPauseAtRound:
			default:
		}
	}
}
```

This way we don't need two setters.

Reply:
In order to stop at round 1 we might need to modify that helper:
```
func (s *Service) updatePauseAtRoundOrBlock() {
	// check for pending pause at round.
	select {
		case s.pauseAtRound = <- s.chanPauseAtRound:
		default:
	}

	if s.pauseAtRound != uint64(0) && uint64(nextRound) > s.pauseAtRound {
		// if we have reached the desired round, block until the pause round changes
		s.pauseAtRound = <-s.chanPauseAtRound
	}
}
```

Reply:
I found a way to merge the earlier two functions
```
/ SetPauseAtRound changes the value of s.pauseAtRound
func (s *Service) SetPauseAtRound(rnd uint64) (err error) {
	// check if pause functionality is enabled
	if !s.dontAllowPauseAtRound {
		// can only pause when the catchup service is running
		if s.pauseAtRound != 0 {
			// only allow resume catchup if either rnd is 0 i.e resume catchup to the latest block
			if rnd == uint64(0) ||
			// check if the catchup has finished reaching the pauseAtRound block number
			// check if next pauseAtRound block number is greater than the latest block on the ledger
				(uint64(s.ledger.LastRound()) >= s.pauseAtRound && rnd > uint64(s.ledger.LastRound())) {
				// send the new value of s.pauseAtRound as the signal and update the value at the other end of the channel to avoid race condition
				s.chanPauseAtRound <- rnd
			} else {
				err = errors.New("not allowed to pause or resume due to an invalid round number")
			}
			// err = errors.New("not allowed to pause when catchup already has a pause round number, try ResumeCatchup")
		} else {
			// takes care of two conditions: when catchup is running as usual and when --round is set before the start
			s.pauseAtRound = rnd
		}
	} else {
		err = errors.New("not allowed to pause")
	}
	return
}
```